DNN Platform/DotNetNuke.Web/Common/RequestScopeServiceProvider.cs

```diff
@@ -0,0 +1,75 @@
+﻿namespace DotNetNuke.Web.Common
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Reflection;
+    using System.Web;
+
+    using DotNetNuke.Common;
+    using DotNetNuke.Common.Extensions;
+    using Microsoft.Extensions.DependencyInjection;
+
+    /// <summary>
+    /// Service provider implementation for <see cref="HttpRuntime.WebObjectActivator"/>.
+    /// </summary>
+    internal class RequestScopeServiceProvider : IServiceProvider
+    {
+        private const int TypesCannotResolveCacheCap = 100000;
+        private readonly ConcurrentDictionary<Type, bool> typesCannotResolve = new ConcurrentDictionary<Type, bool>();
+
+        /// <inheritdoc />
+        public object GetService(Type serviceType)
+        {
+            if (this.typesCannotResolve.ContainsKey(serviceType))
+            {
+                return DefaultCreateInstance(serviceType);
+            }
+
+            var provider = HttpContextSource.Current?.GetScope()?.ServiceProvider;
+
+            if (provider == null)
+            {
+                return DefaultCreateInstance(serviceType);
+            }
+
+            var instance = provider.GetService(serviceType);
+
+            if (instance != null)
+            {
+                return instance;
+            }
+
+            try
+            {
+                return ActivatorUtilities.CreateInstance(provider, serviceType);
+            }
+            catch (InvalidOperationException)
+            {
+                // Thrown by "ActivatorUtilities.CreateInstance" when there is no public constructor available.
+                // Some ASP.NET modules have private constructors, so need to use the default activator instead.
+                if (this.typesCannotResolve.Count < TypesCannotResolveCacheCap)
+                {
+                    this.typesCannotResolve.TryAdd(serviceType, true);
+                }
+
+                return DefaultCreateInstance(serviceType);
+            }
+        }
+
+        /// <summary>
+        /// Create a instance of the given type with <see cref="Activator"/>
+        /// Used when there is no service provider or valid constructor available.
+        /// </summary>
+        /// <param name="serviceType">Requested service type.</param>
+        /// <returns>New instance of the service type.</returns>
+        private static object DefaultCreateInstance(Type serviceType)
+        {
+            return Activator.CreateInstance(
+                serviceType,
+                BindingFlags.Instance | BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.CreateInstance,
+                null,
+                null,
+                null);
+        }
```

Comment:
This method will cause constant crashes in DNN since it is using the `Activator.CreateInstance`. In the scenarios where this would be called, the DI container does not have a registration for the dependency and it tries to resolve it. That dependency may be relying on Dependency Injection or not have a default constructor. At that point it will throw an exception when the `Activator.CreateInstance` is invoked.

I recommend updating this code to use `ActivatorUtilities.GetServiceOrCreateInstance` which will try to resolve it with the container first. If it does not exist on the container it will resolve all the parameters on the constructor with the container and then use the Activator to instantiate it.

Consider this example
```c#
public class MyClass
{
    public MyClass(IHostSettingsService settings)
    {
    }
}
```

if you try injecting the `MyClass` into a WebForms control without registering it, the code would fall through to your `DefaultCreateInstance` method. Even though the class isn't registered it is depending on the `IHostSettingsService` which is registered. The safest way to handle this is using the `ActivatorUtilities.GetServiceOrCreateInstance()`

Reply:
> That dependency may be relying on Dependency Injection or not have a default constructor. At that point it will throw an exception when the Activator.CreateInstance is invoked.

The problem is that not only services are being resolved here. Controls and HttpModules are also being resolved through this service provider. Some of them (for example `System.Web.UI.PageHandlerFactory` which I used in the Unit Test) are being resolved but don't have a public constructor (see https://referencesource.microsoft.com/#System.Web/UI/PageHandlerFactory.cs,31).

If we remove `Activator.CreateInstance` and only use `ActivatorUtilities.GetServiceOrCreateInstance` then these types cannot be resolved or created and will throw a `InvalidOperationException`. Which is invalid.

Reply:
Thank you for giving me a concrete example, it really helped me in understanding the reasoning behind the `Activator.CreateInstance`. Let's finish the conversation in the other thread so we don't have diverging conversations. https://github.com/dnnsoftware/Dnn.Platform/pull/4098#discussion_r491128632